lib/jarvis/clamp_delegate.rb

```diff
@@ -13,19 +13,42 @@ module Jarvis module ClampDelegate
   # The 'foo' command, when invoked via the Lita bot, will pass the whole
   # message body into the Foo clamp command as if it were run via the command
   # line.
-  def self.delegate(command_class)
+  def self.delegate(command_class, options)
     if !command_class.ancestors.include?(Clamp::Command)
       raise ArgumentError, "Cannot delegate to `#{command_class}` because it is not a subclass of Clamp::Command"
     end
-    callback(command_class)
+    callback(command_class, options)
   end
 
   private
-  def self.callback(command_class)
+  def self.callback(command_class, options)
+    if options.include?(:flags) 
+      if !options[:flags].is_a?(Hash) 
+        raise ArgumentError, ":flags option to #{command_class} must be a Hash of { \"--string-flag\" => lambda { |x| ... } }"
+      end
+      if !options[:flags].all? { |k,v| k.is_a?(String) && v.is_a?(Proc) }
+        raise ArgumentError, ":flags option to #{command_class} must be a Hash of { \"--string-flag\" => lambda { |x| ... } }"
+      end
+      if !options[:flags].all? { |k,_| k =~ /^--/ }
+        raise ArgumentError, ":flags keys must all be string starting with '--', like { \"--example\" => lambda { |x| ... } }"
+      end
+    end
+
     lambda do |request|
       name, *args = Shellwords.shellsplit(request.message.body)
       cmd = command_class.new(name)
 
+      if options.include?(:flags)
+        # prepend flags to the args
+        # The `:flags` value is expected to be a Hash of "--flag-name" => lambda { |request| ... }
+        # The idea is that there can be automatic flags based on the
+        # circumstances of the request.
+        options[:flags].each do |key, value_proc|
+          value = value_proc.call(request)
+          args.unshift(key, value) if value
+        end
+      end
+
```

Comment:
I see 4 contexts of execution in this method:
- validate the options
- generating a command
- inject
- execute

I think we could split things up like this, a command class:

Command.new(command_class, options)
Command.valid?
Command.command_method
Command.request

inject_method(cmd.command_method, request)
execute